src/component/I18nRoutingProvider.js

```diff
@@ -0,0 +1,87 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+
+import I18nRoutingContext from './I18nRoutingContext';
+
+class I18nRoutingProvider extends React.Component {
+  state = {
+    locale: null,
+  };
+
+  constructor(props) {
+    super(props);
+
+    const location = props.history ? props.history.location : {};
+
+    this.state = {
+      locale: props.localeFromPath({
+        ...location,
+        hostname: typeof window !== 'undefined' ? window.location.hostname : '',
+      }),
+      translatedRoutes: props.defaultTranslatedRoutes || {},
+      setTranslatedRoutes: this.setTranslatedRoutes.bind(this),
+    };
+  }
+
+  static getDerivedStateFromProps(props, state) {
+    return {
+      ...state,
+      formatIntlRoute: props.formatIntlRoute,
+    };
+  }
+
+  componentDidMount() {
+    if (!this.props.history) {
+      return;
+    }
+
+    this.locationChangeUnlisten = this.props.history.listen(location => {
+      this.handleLocationChange(location);
+    });
+  }
+
+  componentWillUnmount() {
+    if (!this.locationChangeUnlisten) {
+      return;
+    }
+
+    this.locationChangeUnlisten();
+  }
+
+  setTranslatedRoutes(translatedRoutes) {
+    if (translatedRoutes.en === this.state.translatedRoutes.en) {
```

Comment:
It prevents state change when translated routes haven't change, for example in a rerender. The comparation is not very good. We need to deeply compare the two objects

Reply:
@mktoast I improved the check using loadash.isEqual. The downside is that it weighs 11k (4k gziped)

Reply:
We are passing down the `setTranslatedRoutes` as a context property so the developer can update the context's value. Said that, I would not protect a re-render (due to the setState call) since the `setTranslatedRoutes` method call will be explicit.

Reply:
`setTranslatedRoutes` is used typically inside the render method. Therefore unwanted calls may be made

Reply:
https://stackoverflow.com/questions/49809884/access-react-context-outside-of-render-function/49810870#49810870

Reply:
Interesting but that solution is only supported in React >= 16.6.0. Does it worth it?